25_FilterTest/main.cpp

```diff
@@ -1585,78 +1608,23 @@ class BlitFilterTestApp : public ApplicationBase
 		return image;
 	}
 
-	core::smart_refctd_ptr<nbl::ui::IWindowManager> windowManager;
-	core::smart_refctd_ptr<nbl::ui::IWindow> window;
-	core::smart_refctd_ptr<CommonAPI::CommonAPIEventCallback> windowCb;
-	core::smart_refctd_ptr<nbl::video::IAPIConnection> apiConnection;
-	core::smart_refctd_ptr<nbl::video::ISurface> surface;
-	core::smart_refctd_ptr<nbl::video::IUtilities> utilities;
-	core::smart_refctd_ptr<nbl::video::ILogicalDevice> logicalDevice;
-	video::IPhysicalDevice* physicalDevice;
-	std::array<video::IGPUQueue*, CommonAPI::InitOutput::MaxQueuesCount> queues;
-	core::smart_refctd_ptr<nbl::video::ISwapchain> swapchain;
-	core::smart_refctd_ptr<video::IGPURenderpass> renderpass = nullptr;
-	std::array<nbl::core::smart_refctd_ptr<video::IGPUFramebuffer>, CommonAPI::InitOutput::MaxSwapChainImageCount> fbos;
-	std::array<std::array<nbl::core::smart_refctd_ptr<nbl::video::IGPUCommandPool>, CommonAPI::InitOutput::MaxFramesInFlight>, CommonAPI::InitOutput::MaxQueuesCount> commandPools;
-	core::smart_refctd_ptr<nbl::system::ISystem> system;
-	core::smart_refctd_ptr<nbl::asset::IAssetManager> assetManager;
-	video::IGPUObjectFromAssetConverter::SParams cpu2gpuParams;
-	core::smart_refctd_ptr<nbl::system::ILogger> logger;
-	core::smart_refctd_ptr<CommonAPI::InputSystem> inputSystem;
-	video::IGPUObjectFromAssetConverter cpu2gpu;
-
-public:
-	void setWindow(core::smart_refctd_ptr<nbl::ui::IWindow>&& wnd) override
-	{
-		window = std::move(wnd);
-	}
-	void setSystem(core::smart_refctd_ptr<nbl::system::ISystem>&& s) override
-	{
-		system = std::move(s);
-	}
-	nbl::ui::IWindow* getWindow() override
-	{
-		return window.get();
-	}
-	video::IAPIConnection* getAPIConnection() override
-	{
-		return apiConnection.get();
-	}
-	video::ILogicalDevice* getLogicalDevice()  override
+	video::SPhysicalDeviceFeatures getRequiredDeviceFeatures() const override
 	{
-		return logicalDevice.get();
-	}
-	video::IGPURenderpass* getRenderpass() override
-	{
-		return renderpass.get();
-	}
-	void setSurface(core::smart_refctd_ptr<video::ISurface>&& s) override
-	{
-		surface = std::move(s);
-	}
-	void setFBOs(std::vector<core::smart_refctd_ptr<video::IGPUFramebuffer>>& f) override
-	{
-		for (int i = 0; i < f.size(); i++)
-		{
-			fbos[i] = core::smart_refctd_ptr(f[i]);
-		}
-	}
-	void setSwapchain(core::smart_refctd_ptr<video::ISwapchain>&& s) override
-	{
-		swapchain = std::move(s);
-	}
-	uint32_t getSwapchainImageCount() override
-	{
-		return SC_IMG_COUNT;
-	}
-	virtual nbl::asset::E_FORMAT getDepthFormat() override
-	{
-		return nbl::asset::EF_D32_SFLOAT;
+		auto retval = base_t::getRequiredDeviceFeatures();
+
+		retval.vulkanMemoryModelDeviceScope = true;
+
+		return retval;
 	}
 
-	APP_CONSTRUCTOR(BlitFilterTestApp);
+	core::smart_refctd_ptr<asset::IAssetManager> assetManager;
+	video::IGPUQueue* queue;
+	core::smart_refctd_ptr<video::IGPUCommandPool> commandPool;
+	core::smart_refctd_ptr<video::IUtilities> utilities;
```

Comment:
maybe inherit from `BasicMultiQueueApplication` instead of `MonoDeviceApplication` because you use `IUtilities`

due to asset manager usage I also recommend `MonoAssetManagerAndBuiltinResourceApplication` inheritance

Reply:
Actually trying to use `BasicMultiQueueApplication` I realized that the way `IGPUObjectFromAssetConverter` works now it requires both the compute and transfer queue to also support graphics, while `BasicMultiQueueApplication` tries to pick-up the Async-Compute and DMA engines first.

Reply:
> Actually trying to use `BasicMultiQueueApplication` I realized that the way `IGPUObjectFromAssetConverter` works now it requires both the compute and transfer queue to also support graphics, while `BasicMultiQueueApplication` tries to pick-up the Async-Compute and DMA engines first.

yes its a known bug, BECAUSE our asset converter is not using the compute queue to do mip-map generation :)

Next thing on your TODO will be to make it use the HLSL Blit instead, then there will no longer be a requirement for a graphics capable queue from `IGPUObjectFromAssetConverter`

P.S. IIRC its only the compute queue that needs to support graphics.